test/beginend-dired-test.el

```diff
@@ -33,80 +33,110 @@
+  "Test dired.
+FN is the function to be executed.
+EXPECTED is the expected filename.  nil means the end of the buffer."
+  (goto-char (point-min))
+  (funcall fn)
+  (if expected
+      (expect (looking-at expected) :to-be-truthy)
+    (expect (eobp) :to-be-truthy)))
+
+(defun beginend-dired-test-begin (expected)
+  "Test dired goto the beginning.
+EXPECTED is the expected filename.  nil means the end of the buffer."
+  (beginend-dired-test 'beginend-dired-mode-goto-beginning expected))
+
+(defun beginend-dired-test-end (expected)
+  "Test dired goto the end.
+EXPECTED is the expected filename.  nil means the end of the buffer."
+  (beginend-dired-test 'beginend-dired-mode-goto-end expected))
+
+(defmacro with-beginend-dired-test-buffer (specs entries &rest body)
+  "Execute the forms in BODY within a temporary directory.
+The directory is created by SPECS and ENTRIES.  See the documentation of
+`beginend-create-dired-buffer'."
+  (declare (indent 2))
+  `(with-current-buffer (beginend-create-dired-buffer ',specs ,entries)
+     ,@body))
 
 (describe "beginend in a dired buffer"
-  (describe "without dired-hide-details-mode"
-    (describe "without dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p))
-
-      (it "offers a command to go to last file"
-        (beginend-dired-test)))
-
-    (describe "with dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p 'omit))
-
-      (it "offers a command to go to last file with dired-omit-mode"
-        (beginend-dired-test 'omit))))
-
-  (describe "with-dired-hide-details-mode (also hides information lines)"
-    (describe "without dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p 'hide-details))
-
-      (it "offers a command to go to last file with dired-omit-mode"
-        (beginend-dired-test 'hide-details)))
-
-    (describe "with dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p 'omit 'hide-details))
-
-      (it "offers a command to go to last file with dired-omit-mode"
-        (beginend-dired-test 'omit 'hide-details))))
-
-  (describe "with-dired-hide-details-mode (does not hide information lines)"
-    (describe "without dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p 'hide-details 'keep-info-lines))
-
-      (it "offers a command to go to last file with dired-omit-mode"
-        (beginend-dired-test 'hide-details 'keep-info-lines)))
-
-    (describe "with dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p 'omit 'hide-details 'keep-info-lines))
-
-      (it "offers a command to go to last file with dired-omit-mode"
-        (beginend-dired-test 'omit 'hide-details 'keep-info-lines)))))
+  (it "tests . and .. at the beginning"
```

Comment:
please phrase each spec so it forms a sentence with the describe. For example:

```emacs
(it "ignores . and .. at the beginning"
  ...)
```

The full sentence is then "beginend in a dired buffer ignores . and .. at the beginning".